File: BufferLab/buflab-handout/solve.py
import os

def nextrand()

def check(cookie):
	if not(cookie >> 28): return False
	if cookie == 10: return False
	x = 8
	while (cookie >> x) != 10:
		x += 8
		if x == 32:
			return True
	return False

def hash(string):
	result = 0
	for i in range(len(string)):
		result = ord(string[i]) + 103 * result
	return result

def gencookie(username):
	seed = hash(username)
	rand = LCG(seed)
	cookie = rand.next()
	while not check(cookie):
		cookie = rand.next()
	return cookie


print "0x%08x" % gencookie("lol")      # 0x3cc6936f
print "0x%08x" % gencookie("zhan4854") # 0x5245f0fd

File: BufferLab/buflab-handout/solve.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int check(int cookie) {
	if (cookie >> 28) {
		if ((unsigned) cookie == 10) return 0;
		int x = 8;
		while ((unsigned) (cookie >> x) != 10) {
			x += 8;
			if (x == 32) return 1;
		}
		return 0;
	} else return 0;
}

int hash(char* string) {
	int result = 0;
	for (int i = 0; i < strlen(string); ++i) {
		result = string[i] + 103 * result;
	}
	return result;
}

int gencookie(char* username) {
	int seed = hash(username);
	int cookie;
	srand(seed);
	do {
		cookie = rand();
	} while (!check(cookie));
	return cookie;
}

int main() {
	printf("0x%08x\n", gencookie("lol"));
	printf("0x%08x\n", gencookie("zhan4854"));
	return 0;
}